File: binding/src/libqmainwindow.cpp
#include <QByteArray>
#include <QContextMenuEvent>
#include <QDockWidget>
#include <QEvent>
#include <QMainWindow>
#include <QMenu>
#include <QMenuBar>
#include <QPoint>
#include <QSize>
#include <QStatusBar>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QToolBar>
#include <QWidget>
#include <qmainwindow.h>
#include "libqmainwindow.h"
#include "libqmainwindow.hxx"

QMainWindow* QMainWindow_new(QWidget* parent) {
	 return new VirtualQMainWindow(parent);
}

QMainWindow* QMainWindow_new2() {
	 return new VirtualQMainWindow();
}

QMainWindow* QMainWindow_new3(QWidget* parent, int flags) {
	 return new VirtualQMainWindow(parent, static_cast<QFlags<Qt::WindowType>>(flags));
}

libqt_string QMainWindow_Tr(const char* s) {
	QString _ret = QMainWindow::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QSize* QMainWindow_IconSize(const QMainWindow* self) {
	return new QSize(self->iconSize());
}

void QMainWindow_SetIconSize(QMainWindow* self, const QSize* iconSize) {
	self->setIconSize(*iconSize);
}

int QMainWindow_ToolButtonStyle(const QMainWindow* self) {
	return self->toolButtonStyle();
}

void QMainWindow_SetToolButtonStyle(QMainWindow* self, int toolButtonStyle) {
	self->setToolButtonStyle(static_cast<Qt::ToolButtonStyle>(toolButtonStyle));
}

bool QMainWindow_IsAnimated(const QMainWindow* self) {
	return self->isAnimated();
}

bool QMainWindow_IsDockNestingEnabled(const QMainWindow* self) {
	return self->isDockNestingEnabled();
}

bool QMainWindow_DocumentMode(const QMainWindow* self) {
	return self->documentMode();
}

void QMainWindow_SetDocumentMode(QMainWindow* self, bool enabled) {
	self->setDocumentMode(enabled);
}

int QMainWindow_TabShape(const QMainWindow* self) {
	return self->tabShape();
}

void QMainWindow_SetTabShape(QMainWindow* self, int tabShape) {
	self->setTabShape(static_cast<QTabWidget::TabShape>(tabShape));
}

int QMainWindow_TabPosition(const QMainWindow* self, int area) {
	return self->tabPosition(static_cast<Qt::DockWidgetArea>(area));
}

void QMainWindow_SetTabPosition(QMainWindow* self, int areas, int tabPosition) {
	self->setTabPosition(static_cast<QFlags<Qt::DockWidgetArea>>(areas), static_cast<QTabWidget::TabPosition>(tabPosition));
}

void QMainWindow_SetDockOptions(QMainWindow* self, int options) {
	self->setDockOptions(static_cast<QFlags<QMainWindow::DockOption>>(options));
}

int QMainWindow_DockOptions(const QMainWindow* self) {
	return self->dockOptions();
}

bool QMainWindow_IsSeparator(const QMainWindow* self, const QPoint* pos) {
	return self->isSeparator(*pos);
}

QMenuBar* QMainWindow_MenuBar(const QMainWindow* self) {
	return self->menuBar();
}

void QMainWindow_SetMenuBar(QMainWindow* self, QMenuBar* menubar) {
	self->setMenuBar(menubar);
}

QWidget* QMainWindow_MenuWidget(const QMainWindow* self) {
	return self->menuWidget();
}

void QMainWindow_SetMenuWidget(QMainWindow* self, QWidget* menubar) {
	self->setMenuWidget(menubar);
}

QStatusBar* QMainWindow_StatusBar(const QMainWindow* self) {
	return self->statusBar();
}

void QMainWindow_SetStatusBar(QMainWindow* self, QStatusBar* statusbar) {
	self->setStatusBar(statusbar);
}

QWidget* QMainWindow_CentralWidget(const QMainWindow* self) {
	return self->centralWidget();
}

void QMainWindow_SetCentralWidget(QMainWindow* self, QWidget* widget) {
	self->setCentralWidget(widget);
}

QWidget* QMainWindow_TakeCentralWidget(QMainWindow* self) {
	return self->takeCentralWidget();
}

void QMainWindow_SetCorner(QMainWindow* self, int corner, int area) {
	self->setCorner(static_cast<Qt::Corner>(corner), static_cast<Qt::DockWidgetArea>(area));
}

int QMainWindow_Corner(const QMainWindow* self, int corner) {
	return self->corner(static_cast<Qt::Corner>(corner));
}

void QMainWindow_AddToolBarBreak(QMainWindow* self) {
	self->addToolBarBreak();
}

void QMainWindow_InsertToolBarBreak(QMainWindow* self, QToolBar* before) {
	self->insertToolBarBreak(before);
}

void QMainWindow_AddToolBar(QMainWindow* self, int area, QToolBar* toolbar) {
	self->addToolBar(static_cast<Qt::ToolBarArea>(area), toolbar);
}

void QMainWindow_AddToolBar2(QMainWindow* self, QToolBar* toolbar) {
	self->addToolBar(toolbar);
}

QToolBar* QMainWindow_AddToolBar3(QMainWindow* self, const libqt_string title) {
	return self->addToolBar(QString::fromUtf8(title.data, title.len));
}

void QMainWindow_InsertToolBar(QMainWindow* self, QToolBar* before, QToolBar* toolbar) {
	self->insertToolBar(before, toolbar);
}

void QMainWindow_RemoveToolBar(QMainWindow* self, QToolBar* toolbar) {
	self->removeToolBar(toolbar);
}

void QMainWindow_RemoveToolBarBreak(QMainWindow* self, QToolBar* before) {
	self->removeToolBarBreak(before);
}

bool QMainWindow_UnifiedTitleAndToolBarOnMac(const QMainWindow* self) {
	return self->unifiedTitleAndToolBarOnMac();
}

int QMainWindow_ToolBarArea(const QMainWindow* self, const QToolBar* toolbar) {
	return self->toolBarArea(toolbar);
}

bool QMainWindow_ToolBarBreak(const QMainWindow* self, QToolBar* toolbar) {
	return self->toolBarBreak(toolbar);
}

void QMainWindow_AddDockWidget(QMainWindow* self, int area, QDockWidget* dockwidget) {
	self->addDockWidget(static_cast<Qt::DockWidgetArea>(area), dockwidget);
}

void QMainWindow_AddDockWidget2(QMainWindow* self, int area, QDockWidget* dockwidget, int orientation) {
	self->addDockWidget(static_cast<Qt::DockWidgetArea>(area), dockwidget, static_cast<Qt::Orientation>(orientation));
}

void QMainWindow_SplitDockWidget(QMainWindow* self, QDockWidget* after, QDockWidget* dockwidget, int orientation) {
	self->splitDockWidget(after, dockwidget, static_cast<Qt::Orientation>(orientation));
}

void QMainWindow_TabifyDockWidget(QMainWindow* self, QDockWidget* first, QDockWidget* second) {
	self->tabifyDockWidget(first, second);
}

libqt_list QMainWindow_TabifiedDockWidgets(const QMainWindow* self, QDockWidget* dockwidget) {
	return self->tabifiedDockWidgets(dockwidget);
}

void QMainWindow_RemoveDockWidget(QMainWindow* self, QDockWidget* dockwidget) {
	self->removeDockWidget(dockwidget);
}

bool QMainWindow_RestoreDockWidget(QMainWindow* self, QDockWidget* dockwidget) {
	return self->restoreDockWidget(dockwidget);
}

int QMainWindow_DockWidgetArea(const QMainWindow* self, QDockWidget* dockwidget) {
	return self->dockWidgetArea(dockwidget);
}

void QMainWindow_ResizeDocks(QMainWindow* self, const libqt_list docks, const libqt_list sizes, int orientation) {
	self->resizeDocks(*docks, *sizes, static_cast<Qt::Orientation>(orientation));
}

libqt_string QMainWindow_SaveState(const QMainWindow* self) {
	QByteArray _qb = self->saveState();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QMainWindow_RestoreState(QMainWindow* self, const libqt_string state) {
	return self->restoreState(QByteArray(state.data, state.len));
}

QMenu* QMainWindow_CreatePopupMenu(QMainWindow* self) {
	return self->createPopupMenu();
}

void QMainWindow_SetAnimated(QMainWindow* self, bool enabled) {
	self->setAnimated(enabled);
}

void QMainWindow_SetDockNestingEnabled(QMainWindow* self, bool enabled) {
	self->setDockNestingEnabled(enabled);
}

void QMainWindow_SetUnifiedTitleAndToolBarOnMac(QMainWindow* self, bool set) {
	self->setUnifiedTitleAndToolBarOnMac(set);
}

void QMainWindow_IconSizeChanged(QMainWindow* self, const QSize* iconSize) {
	self->iconSizeChanged(*iconSize);
}

void QMainWindow_Connect_IconSizeChanged(QMainWindow* self, intptr_t slot) {
    void (*slotFunc)(QMainWindow*, const QSize*) = reinterpret_cast<void (*)(QMainWindow*, const QSize*)>(slot);
    QMainWindow::connect(self, &QMainWindow::iconSizeChanged, [self, slotFunc](const QSize& iconSize) {
	slotFunc(self, iconSize);
    });
}

void QMainWindow_ToolButtonStyleChanged(QMainWindow* self, int toolButtonStyle) {
	self->toolButtonStyleChanged(static_cast<Qt::ToolButtonStyle>(toolButtonStyle));
}

void QMainWindow_Connect_ToolButtonStyleChanged(QMainWindow* self, intptr_t slot) {
    void (*slotFunc)(QMainWindow*, int) = reinterpret_cast<void (*)(QMainWindow*, int)>(slot);
    QMainWindow::connect(self, &QMainWindow::toolButtonStyleChanged, [self, slotFunc](Qt::ToolButtonStyle toolButtonStyle) {
	slotFunc(self, toolButtonStyle);
    });
}

void QMainWindow_TabifiedDockWidgetActivated(QMainWindow* self, QDockWidget* dockWidget) {
	self->tabifiedDockWidgetActivated(dockWidget);
}

void QMainWindow_Connect_TabifiedDockWidgetActivated(QMainWindow* self, intptr_t slot) {
    void (*slotFunc)(QMainWindow*, QDockWidget*) = reinterpret_cast<void (*)(QMainWindow*, QDockWidget*)>(slot);
    QMainWindow::connect(self, &QMainWindow::tabifiedDockWidgetActivated, [self, slotFunc](QDockWidget* dockWidget) {
	slotFunc(self, dockWidget);
    });
}

libqt_string QMainWindow_Tr2(const char* s, const char* c) {
	QString _ret = QMainWindow::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QMainWindow_Tr3(const char* s, const char* c, int n) {
	QString _ret = QMainWindow::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QMainWindow_AddToolBarBreak1(QMainWindow* self, int area) {
	self->addToolBarBreak(static_cast<Qt::ToolBarArea>(area));
}

libqt_string QMainWindow_SaveState1(const QMainWindow* self, int version) {
	QByteArray _qb = self->saveState(version);
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QMainWindow_RestoreState2(QMainWindow* self, const libqt_string state, int version) {
	return self->restoreState(QByteArray(state.data, state.len), version);
}

// Base class handler implementation
QMenu* QMainWindow_QBaseCreatePopupMenu(QMainWindow* self) {
	auto* vqmainwindow = dynamic_cast<VirtualQMainWindow*>(self);
	if (vqmainwindow && vqmainwindow->isVirtualQMainWindow) {
vqmainwindow->setQMainWindow_CreatePopupMenu_IsBase(true);
	return vqmainwindow->createPopupMenu();
}
}

// Auxiliary method to allow providing re-implementation
void QMainWindow_OnCreatePopupMenu(QMainWindow* self, intptr_t slot) {
	auto* vqmainwindow = dynamic_cast<VirtualQMainWindow*>(self);
	if (vqmainwindow && vqmainwindow->isVirtualQMainWindow) {
vqmainwindow->setQMainWindow_CreatePopupMenu_Callback(reinterpret_cast<VirtualQMainWindow::QMainWindow_CreatePopupMenu_Callback>(slot));
}
}

// Derived class handler implementation
void QMainWindow_ContextMenuEvent(QMainWindow* self, QContextMenuEvent* event) {
	auto* vqmainwindow = dynamic_cast<VirtualQMainWindow*>(self);
	if (vqmainwindow && vqmainwindow->isVirtualQMainWindow) {
	vqmainwindow->contextMenuEvent(event);
	} else {
	self->QMainWindow::contextMenuEvent(event);
}
}

// Base class handler implementation
void QMainWindow_QBaseContextMenuEvent(QMainWindow* self, QContextMenuEvent* event) {
	auto* vqmainwindow = dynamic_cast<VirtualQMainWindow*>(self);
	if (vqmainwindow && vqmainwindow->isVirtualQMainWindow) {
vqmainwindow->setQMainWindow_ContextMenuEvent_IsBase(true);
	vqmainwindow->contextMenuEvent(event);
}
}

// Auxiliary method to allow providing re-implementation
void QMainWindow_OnContextMenuEvent(QMainWindow* self, intptr_t slot) {
	auto* vqmainwindow = dynamic_cast<VirtualQMainWindow*>(self);
	if (vqmainwindow && vqmainwindow->isVirtualQMainWindow) {
vqmainwindow->setQMainWindow_ContextMenuEvent_Callback(reinterpret_cast<VirtualQMainWindow::QMainWindow_ContextMenuEvent_Callback>(slot));
}
}

// Derived class handler implementation
bool QMainWindow_Event(QMainWindow* self, QEvent* event) {
	auto* vqmainwindow = dynamic_cast<VirtualQMainWindow*>(self);
	if (vqmainwindow && vqmainwindow->isVirtualQMainWindow) {
	return vqmainwindow->event(event);
	} else {
	return self->QMainWindow::event(event);
}
}

// Base class handler implementation
bool QMainWindow_QBaseEvent(QMainWindow* self, QEvent* event) {
	auto* vqmainwindow = dynamic_cast<VirtualQMainWindow*>(self);
	if (vqmainwindow && vqmainwindow->isVirtualQMainWindow) {
vqmainwindow->setQMainWindow_Event_IsBase(true);
	return vqmainwindow->event(event);
}
}

// Auxiliary method to allow providing re-implementation
void QMainWindow_OnEvent(QMainWindow* self, intptr_t slot) {
	auto* vqmainwindow = dynamic_cast<VirtualQMainWindow*>(self);
	if (vqmainwindow && vqmainwindow->isVirtualQMainWindow) {
vqmainwindow->setQMainWindow_Event_Callback(reinterpret_cast<VirtualQMainWindow::QMainWindow_Event_Callback>(slot));
}
}

void QMainWindow_Delete(QMainWindow* self) {
    delete self;
}

File: binding/include/libqmainwindow.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qmainwindow.html
pub const qmainwindow = struct {

    /// New constructs a new QMainWindow object.
    pub fn New(parent: ?*anyopaque) QtC.QMainWindow {
        return qtc.QMainWindow_new(@ptrCast(parent));
    }


    /// New2 constructs a new QMainWindow object.
    pub fn New2() QtC.QMainWindow {
        return qtc.QMainWindow_new2();
    }


    /// New3 constructs a new QMainWindow object.
    pub fn New3(parent: ?*anyopaque, flags: i64) QtC.QMainWindow {
        return qtc.QMainWindow_new3(@ptrCast(parent), @intCast(flags));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMainWindow_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmainwindow.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn IconSize(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QMainWindow_IconSize(@ptrCast(self));
    }

    pub fn SetIconSize(self: ?*anyopaque, iconSize: ?*anyopaque) void {
        qtc.QMainWindow_SetIconSize(@ptrCast(self), @ptrCast(iconSize));
    }

    pub fn ToolButtonStyle(self: ?*anyopaque, ) i32 {
        return qtc.QMainWindow_ToolButtonStyle(@ptrCast(self));
    }

    pub fn SetToolButtonStyle(self: ?*anyopaque, toolButtonStyle: i32) void {
        qtc.QMainWindow_SetToolButtonStyle(@ptrCast(self), @intCast(toolButtonStyle));
    }

    pub fn IsAnimated(self: ?*anyopaque, ) bool {
        return qtc.QMainWindow_IsAnimated(@ptrCast(self));
    }

    pub fn IsDockNestingEnabled(self: ?*anyopaque, ) bool {
        return qtc.QMainWindow_IsDockNestingEnabled(@ptrCast(self));
    }

    pub fn DocumentMode(self: ?*anyopaque, ) bool {
        return qtc.QMainWindow_DocumentMode(@ptrCast(self));
    }

    pub fn SetDocumentMode(self: ?*anyopaque, enabled: bool) void {
        qtc.QMainWindow_SetDocumentMode(@ptrCast(self), enabled);
    }

    pub fn TabShape(self: ?*anyopaque, ) i32 {
        return qtc.QMainWindow_TabShape(@ptrCast(self));
    }

    pub fn SetTabShape(self: ?*anyopaque, tabShape: i32) void {
        qtc.QMainWindow_SetTabShape(@ptrCast(self), @intCast(tabShape));
    }

    pub fn TabPosition(self: ?*anyopaque, area: i32) i32 {
        return qtc.QMainWindow_TabPosition(@ptrCast(self), @intCast(area));
    }

    pub fn SetTabPosition(self: ?*anyopaque, areas: i32, tabPosition: i32) void {
        qtc.QMainWindow_SetTabPosition(@ptrCast(self), @intCast(areas), @intCast(tabPosition));
    }

    pub fn SetDockOptions(self: ?*anyopaque, options: i32) void {
        qtc.QMainWindow_SetDockOptions(@ptrCast(self), @intCast(options));
    }

    pub fn DockOptions(self: ?*anyopaque, ) i32 {
        return qtc.QMainWindow_DockOptions(@ptrCast(self));
    }

    pub fn IsSeparator(self: ?*anyopaque, pos: ?*anyopaque) bool {
        return qtc.QMainWindow_IsSeparator(@ptrCast(self), @ptrCast(pos));
    }

    pub fn MenuBar(self: ?*anyopaque, ) QtC.QMenuBar {
        return qtc.QMainWindow_MenuBar(@ptrCast(self));
    }

    pub fn SetMenuBar(self: ?*anyopaque, menubar: ?*anyopaque) void {
        qtc.QMainWindow_SetMenuBar(@ptrCast(self), @ptrCast(menubar));
    }

    pub fn MenuWidget(self: ?*anyopaque, ) QtC.QWidget {
        return qtc.QMainWindow_MenuWidget(@ptrCast(self));
    }

    pub fn SetMenuWidget(self: ?*anyopaque, menubar: ?*anyopaque) void {
        qtc.QMainWindow_SetMenuWidget(@ptrCast(self), @ptrCast(menubar));
    }

    pub fn StatusBar(self: ?*anyopaque, ) QtC.QStatusBar {
        return qtc.QMainWindow_StatusBar(@ptrCast(self));
    }

    pub fn SetStatusBar(self: ?*anyopaque, statusbar: ?*anyopaque) void {
        qtc.QMainWindow_SetStatusBar(@ptrCast(self), @ptrCast(statusbar));
    }

    pub fn CentralWidget(self: ?*anyopaque, ) QtC.QWidget {
        return qtc.QMainWindow_CentralWidget(@ptrCast(self));
    }

    pub fn SetCentralWidget(self: ?*anyopaque, widget: ?*anyopaque) void {
        qtc.QMainWindow_SetCentralWidget(@ptrCast(self), @ptrCast(widget));
    }

    pub fn TakeCentralWidget(self: ?*anyopaque, ) QtC.QWidget {
        return qtc.QMainWindow_TakeCentralWidget(@ptrCast(self));
    }

    pub fn SetCorner(self: ?*anyopaque, corner: i32, area: i32) void {
        qtc.QMainWindow_SetCorner(@ptrCast(self), @intCast(corner), @intCast(area));
    }

    pub fn Corner(self: ?*anyopaque, corner: i32) i32 {
        return qtc.QMainWindow_Corner(@ptrCast(self), @intCast(corner));
    }

    pub fn AddToolBarBreak(self: ?*anyopaque, ) void {
        qtc.QMainWindow_AddToolBarBreak(@ptrCast(self));
    }

    pub fn InsertToolBarBreak(self: ?*anyopaque, before: ?*anyopaque) void {
        qtc.QMainWindow_InsertToolBarBreak(@ptrCast(self), @ptrCast(before));
    }

    pub fn AddToolBar(self: ?*anyopaque, area: i32, toolbar: ?*anyopaque) void {
        qtc.QMainWindow_AddToolBar(@ptrCast(self), @intCast(area), @ptrCast(toolbar));
    }

    pub fn AddToolBar2(self: ?*anyopaque, toolbar: ?*anyopaque) void {
        qtc.QMainWindow_AddToolBar2(@ptrCast(self), @ptrCast(toolbar));
    }

    pub fn AddToolBar3(self: ?*anyopaque, title: []const u8) QtC.QToolBar {
        const title_str = qtc.libqt_string{
    .len = title.len,
    .data = title.ptr,
};
return qtc.QMainWindow_AddToolBar3(@ptrCast(self), title_str);
    }

    pub fn InsertToolBar(self: ?*anyopaque, before: ?*anyopaque, toolbar: ?*anyopaque) void {
        qtc.QMainWindow_InsertToolBar(@ptrCast(self), @ptrCast(before), @ptrCast(toolbar));
    }

    pub fn RemoveToolBar(self: ?*anyopaque, toolbar: ?*anyopaque) void {
        qtc.QMainWindow_RemoveToolBar(@ptrCast(self), @ptrCast(toolbar));
    }

    pub fn RemoveToolBarBreak(self: ?*anyopaque, before: ?*anyopaque) void {
        qtc.QMainWindow_RemoveToolBarBreak(@ptrCast(self), @ptrCast(before));
    }

    pub fn UnifiedTitleAndToolBarOnMac(self: ?*anyopaque, ) bool {
        return qtc.QMainWindow_UnifiedTitleAndToolBarOnMac(@ptrCast(self));
    }

    pub fn ToolBarArea(self: ?*anyopaque, toolbar: ?*anyopaque) i32 {
        return qtc.QMainWindow_ToolBarArea(@ptrCast(self), @ptrCast(toolbar));
    }

    pub fn ToolBarBreak(self: ?*anyopaque, toolbar: ?*anyopaque) bool {
        return qtc.QMainWindow_ToolBarBreak(@ptrCast(self), @ptrCast(toolbar));
    }

    pub fn AddDockWidget(self: ?*anyopaque, area: i32, dockwidget: ?*anyopaque) void {
        qtc.QMainWindow_AddDockWidget(@ptrCast(self), @intCast(area), @ptrCast(dockwidget));
    }

    pub fn AddDockWidget2(self: ?*anyopaque, area: i32, dockwidget: ?*anyopaque, orientation: i32) void {
        qtc.QMainWindow_AddDockWidget2(@ptrCast(self), @intCast(area), @ptrCast(dockwidget), @intCast(orientation));
    }

    pub fn SplitDockWidget(self: ?*anyopaque, after: ?*anyopaque, dockwidget: ?*anyopaque, orientation: i32) void {
        qtc.QMainWindow_SplitDockWidget(@ptrCast(self), @ptrCast(after), @ptrCast(dockwidget), @intCast(orientation));
    }

    pub fn TabifyDockWidget(self: ?*anyopaque, first: ?*anyopaque, second: ?*anyopaque) void {
        qtc.QMainWindow_TabifyDockWidget(@ptrCast(self), @ptrCast(first), @ptrCast(second));
    }

    pub fn TabifiedDockWidgets(self: ?*anyopaque, dockwidget: ?*anyopaque) []const u8 {
        return qtc.QMainWindow_TabifiedDockWidgets(@ptrCast(self), @ptrCast(dockwidget));
    }

    pub fn RemoveDockWidget(self: ?*anyopaque, dockwidget: ?*anyopaque) void {
        qtc.QMainWindow_RemoveDockWidget(@ptrCast(self), @ptrCast(dockwidget));
    }

    pub fn RestoreDockWidget(self: ?*anyopaque, dockwidget: ?*anyopaque) bool {
        return qtc.QMainWindow_RestoreDockWidget(@ptrCast(self), @ptrCast(dockwidget));
    }

    pub fn DockWidgetArea(self: ?*anyopaque, dockwidget: ?*anyopaque) i32 {
        return qtc.QMainWindow_DockWidgetArea(@ptrCast(self), @ptrCast(dockwidget));
    }

    pub fn ResizeDocks(self: ?*anyopaque, docks: []const u8, sizes: []const u8, orientation: i32) void {
        qtc.QMainWindow_ResizeDocks(@ptrCast(self), @ptrCast(docks), @ptrCast(sizes), @intCast(orientation));
    }

    pub fn SaveState(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QMainWindow_SaveState(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qmainwindow.SaveState: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn RestoreState(self: ?*anyopaque, state: []u8) bool {
        const state_str = qtc.libqt_string{
    .len = state.len,
    .data = state.ptr,
};
return qtc.QMainWindow_RestoreState(@ptrCast(self), state_str);
    }

    pub fn CreatePopupMenu(self: ?*anyopaque, ) QtC.QMenu {
        return qtc.QMainWindow_CreatePopupMenu(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnCreatePopupMenu(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QMenu) void {
        qtc.QMainWindow_OnCreatePopupMenu(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreatePopupMenu(self: ?*anyopaque, ) QtC.QMenu {
        return qtc.QMainWindow_QBaseCreatePopupMenu(@ptrCast(self));
    }

    pub fn SetAnimated(self: ?*anyopaque, enabled: bool) void {
        qtc.QMainWindow_SetAnimated(@ptrCast(self), enabled);
    }

    pub fn SetDockNestingEnabled(self: ?*anyopaque, enabled: bool) void {
        qtc.QMainWindow_SetDockNestingEnabled(@ptrCast(self), enabled);
    }

    pub fn SetUnifiedTitleAndToolBarOnMac(self: ?*anyopaque, set: bool) void {
        qtc.QMainWindow_SetUnifiedTitleAndToolBarOnMac(@ptrCast(self), set);
    }

    pub fn IconSizeChanged(self: ?*anyopaque, iconSize: ?*anyopaque) void {
        qtc.QMainWindow_IconSizeChanged(@ptrCast(self), @ptrCast(iconSize));
    }

    pub fn OnIconSizeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMainWindow_Connect_IconSizeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ToolButtonStyleChanged(self: ?*anyopaque, toolButtonStyle: i32) void {
        qtc.QMainWindow_ToolButtonStyleChanged(@ptrCast(self), @intCast(toolButtonStyle));
    }

    pub fn OnToolButtonStyleChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QMainWindow_Connect_ToolButtonStyleChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn TabifiedDockWidgetActivated(self: ?*anyopaque, dockWidget: ?*anyopaque) void {
        qtc.QMainWindow_TabifiedDockWidgetActivated(@ptrCast(self), @ptrCast(dockWidget));
    }

    pub fn OnTabifiedDockWidgetActivated(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMainWindow_Connect_TabifiedDockWidgetActivated(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ContextMenuEvent(self: ?*anyopaque, event: ?*anyopaque) void {
        qtc.QMainWindow_ContextMenuEvent(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnContextMenuEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QMainWindow_OnContextMenuEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseContextMenuEvent(self: ?*anyopaque, event: ?*anyopaque) void {
        qtc.QMainWindow_QBaseContextMenuEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn Event(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QMainWindow_Event(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QMainWindow_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QMainWindow_QBaseEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMainWindow_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmainwindow.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMainWindow_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmainwindow.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AddToolBarBreak1(self: ?*anyopaque, area: i32) void {
        qtc.QMainWindow_AddToolBarBreak1(@ptrCast(self), @intCast(area));
    }

    pub fn SaveState1(self: ?*anyopaque, version: i32, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QMainWindow_SaveState1(@ptrCast(self), version);
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qmainwindow.SaveState1: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn RestoreState2(self: ?*anyopaque, state: []u8, version: i32) bool {
        const state_str = qtc.libqt_string{
    .len = state.len,
    .data = state.ptr,
};
return qtc.QMainWindow_RestoreState2(@ptrCast(self), state_str, version);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMainWindow_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/mainwindow.html#types
pub const enums = struct {
    pub const DockOption = enum {
        pub const AnimatedDocks: i32 = 1;
        pub const AllowNestedDocks: i32 = 2;
        pub const AllowTabbedDocks: i32 = 4;
        pub const ForceTabbedDocks: i32 = 8;
        pub const VerticalTabs: i32 = 16;
        pub const GroupedDragging: i32 = 32;
    };

};
